File: crates/gust-core/src/middleware/security.rs
//! Security headers middleware
//!
//! Adds common security headers to responses.

use crate::{Request, Response};
use super::Middleware;

/// Security configuration
#[derive(Clone)]
pub struct SecurityConfig {
    /// Content-Security-Policy
    pub csp: Option<String>,
    /// X-Frame-Options
    pub frame_options: FrameOptions,
    /// X-Content-Type-Options
    pub content_type_options: bool,
    /// X-XSS-Protection
    pub xss_protection: bool,
    /// Strict-Transport-Security
    pub hsts: Option<HstsConfig>,
    /// Referrer-Policy
    pub referrer_policy: Option<String>,
    /// Permissions-Policy
    pub permissions_policy: Option<String>,
    /// Cross-Origin-Opener-Policy
    pub coop: Option<String>,
    /// Cross-Origin-Embedder-Policy
    pub coep: Option<String>,
    /// Cross-Origin-Resource-Policy
    pub corp: Option<String>,
}

/// X-Frame-Options value
#[derive(Clone)]
pub enum FrameOptions {
    Deny,
    SameOrigin,
    AllowFrom(String),
    None,
}

impl FrameOptions {
    pub fn as_header_value(&self) -> Option<String> {
        match self {
            FrameOptions::Deny => Some("DENY".to_string()),
            FrameOptions::SameOrigin => Some("SAMEORIGIN".to_string()),
            FrameOptions::AllowFrom(uri) => Some(format!("ALLOW-FROM {}", uri)),
            FrameOptions::None => None,
        }
    }
}

/// HSTS configuration
#[derive(Clone)]
pub struct HstsConfig {
    pub max_age: u64,
    pub include_subdomains: bool,
    pub preload: bool,
}

impl Default for HstsConfig {
    fn default() -> Self {
        Self {
            max_age: 31536000, // 1 year
            include_subdomains: true,
            preload: false,
        }
    }
}

impl HstsConfig {
    pub fn as_header_value(&self) -> String {
        let mut parts = vec![format!("max-age={}", self.max_age)];
        if self.include_subdomains {
            parts.push("includeSubDomains".to_string());
        }
        if self.preload {
            parts.push("preload".to_string());
        }
        parts.join("; ")
    }
}

impl Default for SecurityConfig {
    fn default() -> Self {
        Self {
            csp: None,
            frame_options: FrameOptions::Deny,
            content_type_options: true,
            xss_protection: true,
            hsts: None,
            referrer_policy: Some("strict-origin-when-cross-origin".to_string()),
            permissions_policy: None,
            coop: None,
            coep: None,
            corp: None,
        }
    }
}

impl SecurityConfig {
    pub fn new() -> Self {
        Self::default()
    }

    /// Strict security preset
    pub fn strict() -> Self {
        Self {
            csp: Some("default-src 'self'; script-src 'self'; style-src 'self' 'unsafe-inline'; img-src 'self' data:; font-src 'self'; object-src 'none'; frame-ancestors 'none'".to_string()),
            frame_options: FrameOptions::Deny,
            content_type_options: true,
            xss_protection: true,
            hsts: Some(HstsConfig {
                max_age: 31536000,
                include_subdomains: true,
                preload: true,
            }),
            referrer_policy: Some("strict-origin-when-cross-origin".to_string()),
            permissions_policy: Some("geolocation=(), microphone=(), camera=()".to_string()),
            coop: Some("same-origin".to_string()),
            coep: Some("require-corp".to_string()),
            corp: Some("same-origin".to_string()),
        }
    }

    /// API security preset (no CSP, relaxed frame options)
    pub fn api() -> Self {
        Self {
            csp: None,
            frame_options: FrameOptions::None,
            content_type_options: true,
            xss_protection: false, // Not needed for APIs
            hsts: Some(HstsConfig::default()),
            referrer_policy: Some("no-referrer".to_string()),
            permissions_policy: None,
            coop: None,
            coep: None,
            corp: None,
        }
    }

    pub fn csp(mut self, policy: impl Into<String>) -> Self {
        self.csp = Some(policy.into());
        self
    }

    pub fn frame_options(mut self, options: FrameOptions) -> Self {
        self.frame_options = options;
        self
    }

    pub fn hsts(mut self, config: HstsConfig) -> Self {
        self.hsts = Some(config);
        self
    }

    pub fn no_hsts(mut self) -> Self {
        self.hsts = None;
        self
    }

    pub fn referrer_policy(mut self, policy: impl Into<String>) -> Self {
        self.referrer_policy = Some(policy.into());
        self
    }

    pub fn permissions_policy(mut self, policy: impl Into<String>) -> Self {
        self.permissions_policy = Some(policy.into());
        self
    }
}

/// Security middleware
pub struct Security {
    config: SecurityConfig,
}

impl Security {
    pub fn new(config: SecurityConfig) -> Self {
        Self { config }
    }

    pub fn default_headers() -> Self {
        Self::new(SecurityConfig::default())
    }

    pub fn strict() -> Self {
        Self::new(SecurityConfig::strict())
    }

    pub fn api() -> Self {
        Self::new(SecurityConfig::api())
    }
}

impl Default for Security {
    fn default() -> Self {
        Self::default_headers()
    }
}

impl Middleware for Security {
    fn before(&self, _req: &mut Request) -> Option<Response> {
        None
    }

    fn after(&self, _req: &Request, res: &mut Response) {
        // Content-Security-Policy
        if let Some(ref csp) = self.config.csp {
            res.headers.push(("Content-Security-Policy".to_string(), csp.clone()));
        }

        // X-Frame-Options
        if let Some(value) = self.config.frame_options.as_header_value() {
            res.headers.push(("X-Frame-Options".to_string(), value));
        }

        // X-Content-Type-Options
        if self.config.content_type_options {
            res.headers.push(("X-Content-Type-Options".to_string(), "nosniff".to_string()));
        }

        // X-XSS-Protection
        if self.config.xss_protection {
            res.headers.push(("X-XSS-Protection".to_string(), "1; mode=block".to_string()));
        }

        // Strict-Transport-Security
        if let Some(ref hsts) = self.config.hsts {
            res.headers.push(("Strict-Transport-Security".to_string(), hsts.as_header_value()));
        }

        // Referrer-Policy
        if let Some(ref policy) = self.config.referrer_policy {
            res.headers.push(("Referrer-Policy".to_string(), policy.clone()));
        }

        // Permissions-Policy
        if let Some(ref policy) = self.config.permissions_policy {
            res.headers.push(("Permissions-Policy".to_string(), policy.clone()));
        }

        // Cross-Origin-Opener-Policy
        if let Some(ref coop) = self.config.coop {
            res.headers.push(("Cross-Origin-Opener-Policy".to_string(), coop.clone()));
        }

        // Cross-Origin-Embedder-Policy
        if let Some(ref coep) = self.config.coep {
            res.headers.push(("Cross-Origin-Embedder-Policy".to_string(), coep.clone()));
        }

        // Cross-Origin-Resource-Policy
        if let Some(ref corp) = self.config.corp {
            res.headers.push(("Cross-Origin-Resource-Policy".to_string(), corp.clone()));
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_hsts_header() {
        let hsts = HstsConfig {
            max_age: 31536000,
            include_subdomains: true,
            preload: true,
        };

        let value = hsts.as_header_value();
        assert!(value.contains("max-age=31536000"));
        assert!(value.contains("includeSubDomains"));
        assert!(value.contains("preload"));
    }

    #[test]
    fn test_frame_options() {
        assert_eq!(FrameOptions::Deny.as_header_value(), Some("DENY".to_string()));
        assert_eq!(FrameOptions::SameOrigin.as_header_value(), Some("SAMEORIGIN".to_string()));
        assert_eq!(FrameOptions::None.as_header_value(), None);
    }
}

File: packages/app/src/security.ts
/**
 * Security Headers
 * Helmet-style security headers
 */

import type { Handler, ServerResponse } from '@sylphx/gust-core'
import type { Context } from './context'
import type { Middleware } from './types'

export type SecurityOptions = {
	/** Content Security Policy */
	readonly contentSecurityPolicy?: string | false
	/** Cross-Origin-Embedder-Policy */
	readonly crossOriginEmbedderPolicy?: 'require-corp' | 'credentialless' | false
	/** Cross-Origin-Opener-Policy */
	readonly crossOriginOpenerPolicy?:
		| 'same-origin'
		| 'same-origin-allow-popups'
		| 'unsafe-none'
		| false
	/** Cross-Origin-Resource-Policy */
	readonly crossOriginResourcePolicy?: 'same-origin' | 'same-site' | 'cross-origin' | false
	/** Referrer-Policy */
	readonly referrerPolicy?: string | false
	/** Strict-Transport-Security (HSTS) */
	readonly hsts?: { maxAge: number; includeSubDomains?: boolean; preload?: boolean } | false
	/** X-Content-Type-Options */
	readonly noSniff?: boolean
	/** X-DNS-Prefetch-Control */
	readonly dnsPrefetchControl?: boolean
	/** X-Download-Options */
	readonly ieNoOpen?: boolean
	/** X-Frame-Options */
	readonly frameguard?: 'DENY' | 'SAMEORIGIN' | false
	/** X-Permitted-Cross-Domain-Policies */
	readonly permittedCrossDomainPolicies?: 'none' | 'master-only' | 'by-content-type' | 'all' | false
	/** X-XSS-Protection (legacy) */
	readonly xssFilter?: boolean
}

const DEFAULT_CSP = "default-src 'self'"

/**
 * Create security headers middleware
 *
 * Adds security headers to responses (Helmet-style).
 * Works as both global middleware and route-level middleware.
 *
 * @example
 * ```typescript
 * serve({
 *   middleware: security(),
 *   routes: [...]
 * })
 * ```
 */
export const security = (options: SecurityOptions = {}): Middleware => {
	const headers: Record<string, string> = {}

	// Content-Security-Policy
	if (options.contentSecurityPolicy !== false) {
		headers['content-security-policy'] = options.contentSecurityPolicy || DEFAULT_CSP
	}

	// Cross-Origin-Embedder-Policy
	if (options.crossOriginEmbedderPolicy !== false && options.crossOriginEmbedderPolicy) {
		headers['cross-origin-embedder-policy'] = options.crossOriginEmbedderPolicy
	}

	// Cross-Origin-Opener-Policy
	if (options.crossOriginOpenerPolicy !== false) {
		headers['cross-origin-opener-policy'] = options.crossOriginOpenerPolicy || 'same-origin'
	}

	// Cross-Origin-Resource-Policy
	if (options.crossOriginResourcePolicy !== false) {
		headers['cross-origin-resource-policy'] = options.crossOriginResourcePolicy || 'same-origin'
	}

	// Referrer-Policy
	if (options.referrerPolicy !== false) {
		headers['referrer-policy'] = options.referrerPolicy || 'strict-origin-when-cross-origin'
	}

	// Strict-Transport-Security
	if (options.hsts !== false) {
		const hsts = options.hsts || { maxAge: 15552000 } // 180 days
		let hstsValue = `max-age=${hsts.maxAge}`
		if (hsts.includeSubDomains) hstsValue += '; includeSubDomains'
		if (hsts.preload) hstsValue += '; preload'
		headers['strict-transport-security'] = hstsValue
	}

	// X-Content-Type-Options
	if (options.noSniff !== false) {
		headers['x-content-type-options'] = 'nosniff'
	}

	// X-DNS-Prefetch-Control
	if (options.dnsPrefetchControl !== undefined) {
		headers['x-dns-prefetch-control'] = options.dnsPrefetchControl ? 'on' : 'off'
	}

	// X-Download-Options
	if (options.ieNoOpen !== false) {
		headers['x-download-options'] = 'noopen'
	}

	// X-Frame-Options
	if (options.frameguard !== false) {
		headers['x-frame-options'] = options.frameguard || 'SAMEORIGIN'
	}

	// X-Permitted-Cross-Domain-Policies
	if (options.permittedCrossDomainPolicies !== false) {
		headers['x-permitted-cross-domain-policies'] = options.permittedCrossDomainPolicies || 'none'
	}

	// X-XSS-Protection (legacy, but still used)
	if (options.xssFilter !== false) {
		headers['x-xss-protection'] = '0' // Recommended to disable as it can cause vulnerabilities
	}

	return <App>(handler: Handler<Context<App>>): Handler<Context<App>> =>
		async (ctx: Context<App>): Promise<ServerResponse> => {
			const res = await handler(ctx)

			return {
				...res,
				headers: {
					...headers,
					...res.headers, // Allow overrides
				},
			}
		}
}

/**
 * Pre-configured security for strict mode
 */
export const strictSecurity = (): Middleware =>
	security({
		contentSecurityPolicy:
			"default-src 'self'; script-src 'self'; style-src 'self'; img-src 'self' data:; font-src 'self'; object-src 'none'; frame-ancestors 'none'",
		crossOriginEmbedderPolicy: 'require-corp',
		crossOriginOpenerPolicy: 'same-origin',
		crossOriginResourcePolicy: 'same-origin',
		referrerPolicy: 'no-referrer',
		hsts: { maxAge: 31536000, includeSubDomains: true, preload: true },
		frameguard: 'DENY',
	})

/**
 * Pre-configured security for API servers
 */
export const apiSecurity = (): Middleware =>
	security({
		contentSecurityPolicy: false, // APIs don't need CSP
		crossOriginEmbedderPolicy: false,
		crossOriginOpenerPolicy: false,
		crossOriginResourcePolicy: 'same-site',
		referrerPolicy: 'no-referrer',
		hsts: { maxAge: 31536000 },
		frameguard: 'DENY',
	})
